src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -167,6 +154,57 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
     }
   }
 
+  function tryCommitAllLifeCycles(finishedWork : Fiber) {
+    let effectfulFiber = finishedWork.firstEffect;
+    try {
+      while (effectfulFiber) {
+        if (effectfulFiber.effectTag === Update ||
+            effectfulFiber.effectTag === PlacementAndUpdate) {
+          const current = effectfulFiber.alternate;
+          commitLifeCycles(current, effectfulFiber);
+        }
+        const next = effectfulFiber.nextEffect;
+        effectfulFiber = next;
+      }
+    } catch (err) {
+      // Slow path: we want to issue a componentWillUnmount()
+      // for any component that received a componentDidMount()
+      // but won't end up in the tree because of the error.
+      const failedFiber = effectfulFiber;
+      if (failedFiber) {
+        revertLifeCyclesCommittedSoFar(finishedWork, failedFiber);
+      }
+      throw err;
+    }
+  }
+
+  function revertLifeCyclesCommittedSoFar(finishedWork : Fiber, failedEffect : Fiber) {
+    // Gather effects in an array because this is a rare code path.
+    // We need to call them in the reverse order.
+    const fibersCommittedSoFar = [];
```

Comment:
Would it be possible to reuse the `nextEffect` pointer here since you probably don't need it anymore? In that case, it's just a "reversing a singly linked list" problem.